pyteal/ast/subroutine.py

```diff
@@ -15,18 +15,37 @@
 
 
 class SubroutineDefinition:
+    """Class that lifts TEAL's `subr` opcode into a first-class PyTeal citizen"""
+
     nextSubroutineId = 0
 
     def __init__(
         self,
         implementation: Callable[..., Expr],
         returnType: TealType,
         nameStr: str = None,
+        input_types: List[TealType] = None,
     ) -> None:
+        """
+        Args:
+            implementation: The python function that was used to define the subroutine
+            returnType: the TealType the the subroutine should return
+            nameStr (optional): the name that is used to identify the subroutine.
+                If omitted, the name defaults to the implementation's __name__ attribute
+            input_types (optional): list of TealTypes of the arguments of the subroutine.
```

Comment:
My initial impression is that, while this addition makes sense in the context of this PR, it will quickly become outdated when ABI types are introduced, which don't map have an explicit `TealType`.

A longer term solution would be to look into a generic `Expr` class like this issue describes: https://github.com/algorand/pyteal/issues/178

Reply:
This is an excellent point and a much better user experience/API then the quick and dirty way I'm proposing. Given today's PySDK developments, I think I will have plenty of other prep-work to pave the way for this PR so I propose tabling this issue until next week. Then we can have a powwow and hammer out a unified approach forward.

Reply:
I see that in the ABI work proposed version of [ast/subroutine.py](https://github.com/algorand/pyteal/blob/3fb73746bf6d0a25aad63ddb162bf8c62c279e07/pyteal/ast/subroutine.py#L383) is creating a new class `ABIReturnSubroutineFnWrapper` and hence the changes proposed here do not strictly conflict with our ABI work. So I think the way I proposed could technically work. However, I can see that the it is not the most elegant design, or that we might not want to expose it to developers at large, except for testing purposes. So I will propose several ways of going forward and welcome others well:

1. keep as is (new optional parameter `input_types: Optional[List[Tealtype]]`)
2. allow generics to specify the input types (though I haven't tested to see if it works with `mypy`). I.e. by endowing `Expr` and `ScratchVar` as generics we could then define something like:
```python
@Subroutine(TealType.uint64)
def foo(x: ScratchVar[TealType.bytes], y: Expr[TealType.uint64]):
    ...
```
and then infer the generic type `T_x` of `x` for example via: `T_x = inspect.get_annotations(foo)['x'].__args__[0]`
3. Instead of making `Expr` and `ScratchVar` generic, we could create new types `ExprParam(Generic[T])` and `ScratchVarParam(Generic[T])` which are also allowed as parameter types of `@Subroutines` and use an approach similar to no. 2 
4. If we want to hide this for general purpose we could have a new decorator `@SubroutineTester` which inherits from `@Subroutine` and allows doing either no.'s 1 or 3 without touching `@Subroutine`'s functionality

Reply:
I see that in the ABI work proposed version of [ast/subroutine.py](https://github.com/algorand/pyteal/blob/3fb73746bf6d0a25aad63ddb162bf8c62c279e07/pyteal/ast/subroutine.py#L383) is creating a new class `ABIReturnSubroutineFnWrapper` and hence the changes proposed here do not strictly conflict with our ABI work. So I think the way I proposed could technically work. However, I can see that the it is not the most elegant design, or that we might not want to expose it to developers at large, except for testing purposes. So I will propose several ways of going forward and welcome others well:

1. keep as is (new optional parameter `input_types: Optional[List[Tealtype]]`)
2. allow generics to specify the input types (though I haven't tested to see if it works with `mypy`). I.e. by endowing `Expr` and `ScratchVar` as generics we could then define something like:
```python
@Subroutine(TealType.uint64)
def foo(x: ScratchVar[TealType.bytes], y: Expr[TealType.uint64]):
    ...
```
and then infer the generic type `T_x` of `x` for example via: `T_x = inspect.get_annotations(foo)['x'].__args__[0]`
3. Instead of making `Expr` and `ScratchVar` generic, we could create a new types `SubrParam(Generic[
4. If we want to hide this for general purpose we could have a new decorator `@SubroutineTester` which inherits from `@Subroutine` and allows doing either 1 or 3 without touching `@Subroutine`'s functionality

_BTW, my preference is no. 1, because it's already done and the others seem very complicated_

Reply:
Thanks for expanding on all the choices here. My current thoughts are that 2 is most preferred, but 3 and 4 would be good solutions as well.

Though _if_ we decide this interface should be internal-only, then I'd say 4 is the most compelling, since it requires the least additional work. We can then change our approach over time as we see fit without endangering backwards compatibility of user code.

(I'm still reviewing the rest of this PR, but I thought you'd appreciate this suggestion ASAP)

Reply:
I'm going to experiment with @michaeldiamant 's #5, but just create the `@Blackbox` decorator in `utils/` that way we don't even touch `pyteal/`. Will keep you posted!

Reply:
Please refer to #297 which implements this approach